Cleaned up Probabilistic

Salahub opened the pull request:
A few changes:
- updated indel() with CRISPResso data to produce results consistent with the CRISPResso distribution
- updated the time step probability to follow an exponential distribution in time rather than a uniform one
- changed the concentration probability function so that it asymptotically approaches 1, and is clearly unitless

With these changes there is one important thing in particular to note:
- the new concentration probability function does not fit the data as well as the last one, however it is unitless, will never produce an output greater than one, and still forms a reasonable approximation of the chart in Wu et al; as a result I changed the function to that and left the old, data-fitted function commented in beneath it

Salahub committed:
Added improved indel function and fixed up the concentration probability function

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -49,7 +64,7 @@ def prob_cut(grna, target, concentration, dt):
     """
     assert len(grna) == len(target)
-    prob_factor_time = dt/average_cut_time  # assume uniform distribution in time - TODO change to poisson
+    prob_factor_time = 1 - math.exp(-(average_cut_time)*dt) 
```
this should be `1 - math.exp(-dt / average_cut_time)`

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -1,5 +1,7 @@
-from random import randint
+from numpy.random import randint
+from numpy.random import choice
+import math
```
can we change these lines to:

```
import math
from numpy import prod
from numpy.random import choice, randint
```

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -68,8 +83,13 @@ def nt_rand(insertion_size):
-    del_left = randint(0,4)
-    del_right = randint(0,4)
+    size = choice(indel_sizes, 1, p = indel_probs)
     insert = randint(0,8)
```
why is insert still random?

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -8,6 +10,13 @@
 # average time for SpyCas9 cutting in perfect match case
 average_cut_time = 60.0
 
+# distribution of indel size from CRISPResso
```
delete the extra line (12), I'd like to keep the constants lumped (planning to move them later)

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -68,8 +83,13 @@ def nt_rand(insertion_size):
     insert = randint(0,8)
-
+    total_del = abs(size - insert)[0]
+    if not total_del == 0:
```
`if total_del != 0:`

mattsmart commented on models/targeting/probabilistic.py:
```diff
@@ -68,8 +83,13 @@ def nt_rand(insertion_size):
+    total_del = abs(size - insert)[0]
+    if not total_del == 0:
+        del_left = randint(math.floor(total_del/4), total_del)
+        del_right = total_del - del_left
```
these lines warrant explanation in the function docstring

mattsmart commented:
minor comments, can merge once you've committed the changes

mattsmart commented:
also: not sure why the branch has a conflict, may need to resubmit PR..

Salahub committed:
Added recommended changes

Salahub committed:
Merge branch 'master' of https://github.com/igem-waterloo/uwaterloo-igem-2015

Conflicts:
	models/targeting/probabilistic.py

Salahub committed:
Fixed the messed up bits

Salahub commented:
So I just fixed the conflict and added the recommended changes. Looks like my branch was just a bit behind before. Should be good now. The third commit was that issue getting resolved and the fourth was me fixing some bits of probabilistic that were messed up in the process

mattsmart commented:
LGTM, merging
